lib/spack/spack/builder.py

```diff
@@ -0,0 +1,512 @@
+    """
+
+    def _decorator(cls):
+        cls.build_system = build_system_name
+        BUILDER_CLS[build_system_name] = cls
+        return cls
+
+    return _decorator
+
+
+def create(pkg):
+    """Given a package object with an associated concrete spec,
+    return the builder object that can install it.
+
+    Args:
+         pkg (spack.package_base.PackageBase): package for which we want the builder
+    """
+    if id(pkg) not in _BUILDERS:
+        _BUILDERS[id(pkg)] = _create(pkg)
+    return _BUILDERS[id(pkg)]
+
+
+class _PhaseAdapter(object):
+    def __init__(self, builder, phase_fn):
+        self.builder = builder
+        self.phase_fn = phase_fn
+
+    def __call__(self, spec, prefix):
+        return self.phase_fn(self.builder.pkg, spec, prefix)
+
+
+def _create(pkg):
+    package_module = inspect.getmodule(pkg)
+    package_buildsystem = buildsystem_name(pkg)
+    default_builder_cls = BUILDER_CLS[package_buildsystem]
+    builder_cls_name = default_builder_cls.__name__
+    builder_cls = getattr(package_module, builder_cls_name, None)
+    if builder_cls:
+        return builder_cls(pkg)
+
+    # Specialized version of a given buildsystem can subclass some
+    # base classes and specialize certain phases or methods or attributes.
+    # In that case they can store their builder class as a class level attribute.
+    # See e.g. AspellDictPackage as an example.
+    base_cls = getattr(pkg, builder_cls_name, default_builder_cls)
+
+    # From here on we define classes to construct a special builder that adapts to the
+    # old, single class, package format. The adapter forwards any call or access to an
+    # attribute related to the installation procedure to a package object wrapped in
+    # a class that falls-back on calling the base builder if no override is found on the
+    # package. The semantic should be the same as the method in the base builder were still
+    # present in the base class of the package.
+
+    class _ForwardToBaseBuilder(object):
+        def __init__(self, wrapped_pkg_object, root_builder):
+            self.wrapped_package_object = wrapped_pkg_object
+            self.root_builder = root_builder
+
+            package_cls = type(wrapped_pkg_object)
+            wrapper_cls = type(self)
+            bases = (package_cls, wrapper_cls)
+            new_cls_name = package_cls.__name__ + "Wrapper"
+            new_cls = type(new_cls_name, bases, {})
+            new_cls.__module__ = package_cls.__module__
+            self.__class__ = new_cls
+            self.__dict__.update(wrapped_pkg_object.__dict__)
+
+        def __getattr__(self, item):
+            default_item = getattr(base_cls, item)
+            setattr(type(self.root_builder), item, default_item)
+            result = getattr(self.root_builder, item)
+            if item in base_cls.phases:
+                result = _PhaseAdapter(self.root_builder, result)
+            return result
```

Comment:
> For example these forwarding methods would have to be defined in CMakePackage etc. to support super()

I'd avoid adding those methods to classes that would end up in the `Package` hierarchy since:
1. It would add back functions that have been extracted from the hierarchy (therefore undoing some tidying currently present in this PR)
2. It would distribute code that is only needed to ensure backward compatibility to multiple places (right now it's only in `builder.py`)
3. The functions would be always present, and names might clash in case of inheritance from multiple build-systems. This wrapper on the other hand is limited in scope and constructed based on the build-system being currently used.

Overall, I prefer this approach to distributing the functions statically to `CMakePackage` etc.

Reply:
From today's discussion: we agreed not to pursue my proposal, so I'll mark this as resolved.